Model/Api/ShippingMethods.php

```diff
@@ -526,15 +535,22 @@ private function getShippingOptions($quote, $addressData)
             ////////////////////////////////////////////////////////////////
 
             $shippingAddress->setShippingMethod($method);
+            // in order to get correct shipping discounts
+            // the following method must be called twice
+            $this->totalsCollector->collectAddressTotals($quote, $shippingAddress);
             $this->totalsCollector->collectAddressTotals($quote, $shippingAddress);
 
-            $cost         = $shippingAddress->getShippingAmount();
+            $cost        = $shippingAddress->getShippingAmount() - $shippingAddress->getShippingDiscountAmount();
             $roundedCost = $this->cartHelper->getRoundAmount($cost);
 
             $diff = $cost * 100 - $roundedCost;
 
             $taxAmount = $this->cartHelper->getRoundAmount($shippingAddress->getTaxAmount() + $diff / 100);
 
+            $discount = $this->priceHelper->currency($shippingAddress->getShippingDiscountAmount(), true, false);
+
+            if ($shippingAddress->getShippingDiscountAmount()) $service .= " [$discount" . "_discount]";
```

Comment:
1. Can you remove underscore? eg., "[$5 discount]"
2. If shipping is free after discount, can you instead say " [free shipping discount]"?

Reply:
@daisy1754 Will do as you say. The reasons I did it this way:
1. underscore was put to prevent splitting into two lines
2. free shipping discount is another discount type, this one simply subtracts the discount amount from the shipping, and if the shipping is <= discount it becomes 0, free.
But as I said, will do as requested.